### PR TITLE
Fix crash during artifact download

### LINKED ISSUE #14143: Downloading artifacts fails on Windows
This is on the latest master:

```
type 'MappedListIterable<String, dynamic>' is not a subtype of type 'List<String>' of 'convertedPieces' where
  MappedListIterable is from dart:_internal
  String is from dart:core
  List is from dart:core
  String is from dart:core

#0      _flattenNameSubdirs (package:flutter_tools/src/cache.dart:521)
#1      CachedArtifact._downloadArchive (package:flutter_tools/src/cache.dart:293)
#2      CachedArtifact._downloadZipArchive (package:flutter_tools/src/cache.dart:309)
#3      FlutterEngine.updateInner (package:flutter_tools/src/cache.dart:447)
<asynchronous suspension>
#4      CachedArtifact.update (package:flutter_tools/src/cache.dart:256)
<asynchronous suspension>
#5      Cache.updateAll (package:flutter_tools/src/cache.dart:223)
<asynchronous suspension>
#6      FlutterCommand.verifyThenRunCommand (package:flutter_tools/src/runner/flutter_command.dart:257)
<asynchronous suspension>
#7      FlutterCommand.run (package:flutter_tools/src/runner/flutter_command.dart:211)
<asynchronous suspension>
#8      CommandRunner.runCommand (package:args/command_runner.dart:194)
<asynchronous suspension>
#9      FlutterCommandRunner.runCommand (package:flutter_tools/src/runner/flutter_command_runner.dart:283)
<asynchronous suspension>
#10     CommandRunner.run.<anonymous closure> (package:args/command_runner.dart:109)
#11     new Future.sync (dart:async/future.dart:222)
#12     CommandRunner.run (package:args/command_runner.dart:109)
#13     FlutterCommandRunner.run (package:flutter_tools/src/runner/flutter_command_runner.dart:166)
#14     run.<anonymous closure> (package:flutter_tools/runner.dart:89)
<asynchronous suspension>
#15     AppContext._run (package:flutter_tools/src/base/context.dart:76)
<asynchronous suspension>
#16     AppContext.runInZone.<anonymous closure> (package:flutter_tools/src/base/context.dart:66)
#17     _rootRun (dart:async/zone.dart:1126)
#18     _CustomZone.run (dart:async/zone.dart:1023)
#19     runZoned (dart:async/zone.dart:1501)
#20     AppContext.runInZone (package:flutter_tools/src/base/context.dart:65)
#21     run (package:flutter_tools/runner.dart:60)
<asynchronous suspension>
#22     main (package:flutter_tools/executable.dart:44)
<asynchronous suspension>
#23     main (file:///C:/src/flutter/packages/flutter_tools/bin/flutter_tools.dart:16)
#24     _startIsolate.<anonymous closure> (dart:isolate-patch/dart:isolate/isolate_patch.dart:275)
#25     _RawReceivePortImpl._handleMessage (dart:isolate-patch/dart:isolate/isolate_patch.dart:163)
```

### PR DESCRIPTION
**Pull request by goderbauer**
Fixes https://github.com/flutter/flutter/issues/14143.

I wish the analyzer would catch these :(

/cc @jcollins-g

### CODE REVIEW

File: packages/flutter_tools/test/cache_test.dart [hunk 1]
```diff
@@ -77,6 +77,14 @@ void main() {
       verify(artifact2.update());
     });
   });
+
+  testUsingContext('flattenNameSubdirs', () {
+    expect(flattenNameSubdirs(Uri.parse('http://flutter.io/foo/bar')), 'flutter.io/foo/bar');
```
Comment (jcollins-g) [inline packages/flutter_tools/test/cache_test.dart]: this won't work on windows, but otherwise looks fine

Comment (goderbauer) [reply packages/flutter_tools/test/cache_test.dart]: For this test I am using the MemoryFileSystem (see below), which produces the same paths on Windows and Linux. So, I would expect this to work.

Comment (jcollins-g) [reply packages/flutter_tools/test/cache_test.dart]: Ah, well, that should indeed work then.

What's surprising to me is that given this, it ever worked at all.

**Review by jcollins-g — APPROVED** [review]
LGTM

**Pull request merged**